packages/back-end/src/events/handlers/slack/handleFeatureEventForSlack.ts

```diff
@@ -0,0 +1,377 @@
+import uniq from "lodash/uniq";
+import isEqual from "lodash/isEqual";
+import intersection from "lodash/intersection";
+import { KnownBlock } from "@slack/web-api";
+import {
+  FeatureCreatedNotificationEvent,
+  FeatureDeletedNotificationEvent,
+  FeatureUpdatedNotificationEvent,
+} from "../../base-events";
+import { getSlackIntegrationsForFilters } from "../../../models/SlackIntegrationModel";
+import { FeatureInterface } from "../../../../types/feature";
+import { SlackIntegrationInterface } from "../../../../types/slack-integration";
+import { logger } from "../../../util/logger";
+import { APP_ORIGIN } from "../../../util/secrets";
+import { sendSlackMessage, SlackMessage } from "./slack-event-handler-utils";
+
+type HandleFeatureEventOptions = {
+  organizationId: string;
+  eventId: string;
+  featureEvent:
+    | FeatureCreatedNotificationEvent
+    | FeatureUpdatedNotificationEvent
+    | FeatureDeletedNotificationEvent;
+};
+export const handleFeatureEventForSlack = async ({
+  organizationId,
+  eventId,
+  featureEvent,
+}: HandleFeatureEventOptions) => {
```

Comment:
This is fine for now, but will likely need to be refactored a bit when we add more than just feature events.  The bulk of the code in this file will be common to all (or most) event types - getting slack integrations for an org, filtering by tags/projects, building the slack message context, sending the message, etc..  I don't think having a separate file for every event entity is going to scale well.

Only a few parts will be different for each event - getting the affected tags/projects/environments, building the message text, etc..  We could pretty easily add a single method like `getEventSlackData` that has a switch statement and everywhere else in the code can be agnostic to the event type.  For example:

```ts
const { tags, projects, environments, message, actionText, actionURL } = getEventSlackData(event);

// At this point, you don't care what the event type is, you're just working with strings and arrays of strings
```

I think this general approach will drastically simplify the code and make it easier to maintain in the future.

I'm fine sticking with the current approach for now, just wanted to call this out in case you thought it was worth fixing now vs waiting until we have another event type and then abstracting things.

Reply:
I've made some changes in the commit 1fb1c95f063f797d6614c1dca7b10819e5d3312b to DRY up some of the common functionality, and put everything in a single file rather than splitting each resource into its own file.

I didn't abstract the Slack message shape as suggested because I think it's important to allow flexibility for multiple blocks in the event we want to have different blocks depending on the resource and action. This is especially important when we take on the work to do a real Slack integration (and not the DIY one) where we will have the ability to use button actions (in a cleaner way) and should therefore have the flexibility to build blocks as we see fit for each message. I don't want to back us into a corner where we lose the flexibility to shape each message as we want.

Other than shaping the Slack message, I've abstracted getting the tags, projects, environments from the event in a shared function for each property. Then we get the SlackIntegration records based on the filtering criteria. Then, we filter further for relevance, where each event name can choose whether to omit some updates as irrelevant. The whole flow can now be read in `slackEventHandler`.

Let me know your thoughts.